eventauth.go

```diff
@@ -976,6 +994,58 @@ func (m *membershipAllower) membershipAllowed(event *Event) error { // nolint: g
 	return m.membershipAllowedOther()
 }
 
+func (m *membershipAllower) membershipAllowedForRestrictedJoin() error {
+	// Special case for restricted room joins, where we will check if the membership
+	// event is signed by one of the allowed servers in the join rule content.
+	allowsRestricted, err := m.roomVersion.AllowRestrictedJoinsInEventAuth()
+	if err != nil {
+		return err
+	}
+	if !allowsRestricted {
+		return fmt.Errorf("restricted joins are not supported in this room version")
+	}
+
+	// In the case that the user is already joined, invited or there is no
+	// authorised via server, we should treat the join rule as if it's invite.
+	if m.oldMember.Membership == Join || m.oldMember.Membership == Invite || m.newMember.AuthorisedVia == "" {
+		m.joinRule.JoinRule = Invite
+		return nil
+	}
+
+	// Otherwise, we have to work out if the server that produced the join was
+	// authorised to do so. This requires the membership event to contain a
+	// 'join_authorised_via_users_server' key, containing the user ID of a user
+	// in the room that should have a suitable power level to issue invites.
+	// If no such key is specified then we should reject the join.
+	if _, _, err := SplitID('@', m.newMember.AuthorisedVia); err != nil {
```

Comment:
What's stopping servers from lying about who authorised this? Can I not just be malicious and guess that "hey Alice is probably in $secret_room, let's slap an authorised key as alice and I can get in"?

Reply:
Well, the authorising server has to have signed the event as well. I assume the malicious person could enter a different user ID from the same signing server, but unless that user has the power levels to invite, I don't suppose it'd help them at all.